### ISSUE TITLE
MutatingAdmissionPolicy mutation ordering issue

### ISSUE THREAD

**Issue by kfox1111** [kubernetes/kubernetes#129309]
### What happened?

This works:
```
    - patchType: "JSONPatch"
      jsonPatch:
        expression: >
          [
            JSONPatch{
              op: "add", path: "/spec/initContainers",
              value: []
            },
            JSONPatch{
              op: "add", path: "/spec/initContainers/-",
              value: Object.spec.initContainers{
                name: "mesh-proxy",
                image: "mesh-proxy/v1.0.0",
                restartPolicy: "Always"
              }
            }
          ]
```

But, this fails for a pod:
```
    - patchType: "JSONPatch"
      jsonPatch:
        expression: >
          [
            JSONPatch{
              op: "add", path: "/spec/initContainers",
              value: []
            }
          ]
    - patchType: "JSONPatch"
      jsonPatch:
        expression: >
          [
            JSONPatch{
              op: "add", path: "/spec/initContainers/-",
              value: Object.spec.initContainers{
                name: "mesh-proxy",
                image: "mesh-proxy/v1.0.0",
                restartPolicy: "Always"
              }
            }
          ]
```
with:
```
denied request: JSON Patch: add operation does not apply: doc is missing path: "/spec/initContainers/-": missing value
```

Seems like the ordering is wrong or the input of one isn't fed into the next

### What did you expect to happen?

Both should work

### How can we reproduce it (as minimally and precisely as possible)?

Try chaining the mutations together in the same object, one relying on the next as above

### Kubernetes version

<details>

```console
Client Version: v1.32.0
Kustomize Version: v5.5.0
Server Version: v1.32.0
```

</details>

### Cloud provider

NA

### OS version

minikube

### Install tools

minikube

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by jpbetz** [comment]
@knrc Thanks for reporting this.  Are you interested in proposing a PR to fix or would you prefer someone else resolve this one?

**Comment by jpbetz** [comment]
Looking at this more closely, it's less clear to me that this is the wrong behavior.

The reason the first case works is that we apply the mutations in series without interleaving in any other operations.  This allows the `value: []` to be preserved between operations.

The reason the second case fails is that we reserialize the object (and apply defaulting, I believe) and so the `value: []` is omitted, causing the second patch operation to fail.

What I'm suggesting at is that we might not want to try to guarantee that a single JSON Patch mutations with multiple operations is the same as multiple JSON Patch mutations that perform the same operations, but as different mutations.

I can see how this can be confusing, but the only viable alternative, I think, is to change the behavior so that BOTH examples fail in the same way, and I think that results in JSON Patch being harder to use because intermediate states get wiped.

**Comment by knrc** [comment]
@jpbetz I'm not the reporter, that's @kfox1111, but I did take a quick look at the issue last week.  I suggested he keep this open so that it could be discussed, at the very least to document this is how it behaves.  It's the serialization between thee mutations that's causing the issue, and this is likely unexpected.

**Comment by jpbetz** [comment]
@knrc gotcha. Yes, this might be best handled via documentation.

**Comment by kfox1111** [comment]
Yeah. If its working through all the patches in a row, actually allowing changes to flow from one to the next, just that the omit functionality happens to be undoing things in this example, I think that is probably okish, and there is no bug here.

**Comment by cici37** [comment]
/close

As discussed above, we would improve the documentation for MAP ordering later: https://github.com/kubernetes/website/issues/49641
Thanks for raising it!
